packages/appium/lib/appium.js

```diff
@@ -241,6 +252,330 @@ class AppiumDriver extends DriverCore {
+      // can be multiple sockets per session
+      if (!this.bidiSockets[sessionId]) {
+        this.bidiSockets[sessionId] = [];
+      }
+      this.bidiSockets[sessionId].push(ws);
+
+      const bidiProxyUrl = bidiHandlerDriver.bidiProxyUrl;
+      if (bidiProxyUrl) {
+        if (!_.isString(bidiProxyUrl)) {
+          throw new Error(
+            `Got request to proxy bidi connections to upstream socket with url ` +
+              `${bidiProxyUrl}, but this was not a valid url`,
+          );
+        }
+        this.log.info(`Bidi connection will be proxied to ${bidiProxyUrl}`);
+        proxyClient = new WebSocket(bidiProxyUrl);
+        this.bidiProxyClients[sessionId] = proxyClient;
+      }
+    } else {
+      this.log.info('Bidi websocket connection made to main server');
+      // no need to store the socket connection if it's to the main server since it will just
+      // stay open as long as the server itself is and will close when the server closes.
+      bidiHandlerDriver = this; // eslint-disable-line @typescript-eslint/no-this-alias
+    }
+
+    const logSocketErr = (/** @type {Error} */ err) =>
+      this.log.error(`${outOfBandErrorPrefix}${err}`);
+
+    // This is a function which wraps the 'send' method on a web socket for two reasons:
+    // 1. Make it async-await friendly
+    // 2. Do some logging if there's a send error
+    const sendFactory = (/** @type {import('ws').WebSocket} */ socket) => {
+      const socketSend = B.promisify(socket.send, {context: socket});
+      return async (/** @type {string|Buffer} */ data) => {
+        const err = await socketSend(data);
+        if (err) {
+          logSocketErr(err);
+        }
+      };
+    };
+
+    // Construct our send method for sending messages to the client
+    const send = sendFactory(ws);
+
+    // Construct a conditional send method for proxying messages from the client to an upstream
+    // bidi socket server (e.g. on a browser)
+    let sendToProxy = null;
+    if (proxyClient) {
+      sendToProxy = sendFactory(proxyClient);
+    }
+
+    return {bidiHandlerDriver, proxyClient, send, sendToProxy, logSocketErr};
+  }
+
+  /**
+   * Set up handlers on upstream bidi socket we are proxying to/from
+   *
+   * @param {import('ws').WebSocket | null} proxyClient - the websocket connection to/from the
+   * upstream socket (the one we're proxying to/from)
+   * @param {import('ws').WebSocket} ws - the websocket connection to/from the client
+   * @param {(data: string | Buffer) => Promise<void>} send - a method used to send data to the
+   * client
+   */
+  initBidiProxyHandlers(proxyClient, ws, send) {
+    // Set up handlers for events that might come from the upstream bidi socket connection if
+    // we're in proxy mode
+    if (proxyClient) {
+      // Here we're receiving a message from the upstream socket server. We want to pass it on to
+      // the client
+      proxyClient.on('message', async (/** @type {Buffer} */ data) => {
```

Comment:
I assume data could also be a string or we don't care here?

Reply:
i can update to include string